src/util/metadata/vo_bit/helper.rs

```diff
@@ -0,0 +1,178 @@
+//! This module updates of VO bits during GC.  It is used for spaces that do not clear the metadata
+//! of some dead objects during GC.  Currently, only ImmixSpace is affected.
+//!
+//! | Policy            | When are VO bits of dead objects cleared                      |
+//! |-------------------|---------------------------------------------------------------|
+//! | MarkSweepSpace    | when sweeping cells of dead objects                           |
+//! | MarkCompactSpace  | when compacting                                               |
+//! | CopySpace         | when releasing the space                                      |
+//!
+//! The policies listed above trivially clear the VO bits for dead objects (individually or in
+//! bulk), and make the VO bits available during tracing.
+//!
+//! For ImmixSpace, if a line contains both live and dead objects, live objects will be traced,
+//! but dead objects will not be visited.  Therefore we cannot clear the VO bits of individual
+//! dead objects.  We cannot clear all VO bits for the line in bulk because it contains live
+//! objects.  This module updates the VO bits for such regions (e.g. Immix lines, or Immix blocks
+//! if Immix is configured to be block-only).
+//!
+//! We implement several strategies depending on whether mmtk-core or the VM binding also requires
+//! the VO bits to also be available during tracing.
+//!
+//! The handling is very sensitive to `VOBitUpdateStrategy`, and may be a bit verbose.
+//! We abstract VO-bit-related code out of the main GC algorithms (such as Immix) to make it more
+//! readable.
+
+use atomic::Ordering;
+
+use crate::{
+    util::{linear_scan::Region, metadata::vo_bit, ObjectReference},
+    vm::{ObjectModel, VMBinding},
+};
+
+/// The strategy to update the valid object (VO) bits.
+///
+/// Each stategy has its strength and limitation.  We should choose a strategy according to the
+/// configuration of the VM binding.  See [`strategy`].
+#[derive(Debug)]
+enum VOBitUpdateStrategy {
+    /// Clear all VO bits after stacks are scanned, and reconstruct the VO bits during tracing.
+    ///
+    /// Pros:
+    /// -   Minimum overhead.
+    ///
+    /// Cons:
+    /// -   VO bits are not available during tracing.
+    ///
+    /// This strategy is described in the paper *Fast Conservative Garbage Collection* published
+    /// in OOPSLA'14.  See: <https://dl.acm.org/doi/10.1145/2660193.2660198>
+    ClearAndReconstruct,
+    /// Copy the mark bits metadata over to the VO bits metadata after tracing.
+    ///
+    /// Pros:
+    /// -   VO bits are available during tracing.
+    ///
+    /// Cons:
+    /// -   Requires marking bits to be on the side.
+    /// -   Has extra time overhead.
```

Comment:
This is no longer true, according to the experiments.

Reply:
I updated the comments to reflect our experiment results.